sos/collector/clusters/saltstack.py

```diff
@@ -0,0 +1,107 @@
+# Copyright Red Hat 2022, Trevor Benson <trevor.benson@gmail.com>
+
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+import json
+from shlex import quote
+from sos.collector.clusters import Cluster
+
+
+class saltstack(Cluster):
+    """
+    The saltstack cluster profile is intended to be used on saltstack
+    clusters (Salt Project).
+    """
+
+    cluster_name = "Saltstack"
+    packages = ("salt-master",)
+    sos_plugins = ["saltmaster"]
+    cmd = "salt-run --out=pprint manage.up"
+    hostname_cmd = "salt --out=newline_values_only {minion} grains.get fqdn"
+    strict_node_list = True
+    option_list = [
+        ("all_nodes", False, "Filter node list to all nodes, even if "
+         " salt-mionion service is down."),
+        ("compound", "", "Filter node list to those matching compound"),
+        ("glob", "", "Filter node list to those matching glob pattern"),
+        ("grain", "", "Filter node list to those with matching grain"),
+        ("hostnames", False, "When minion isn't resolvable for control_persist"
+         " this will return the fqdn of the minion. This option is not"
+         " compatible with option all_nodes or transport saltstack"),
+        ("list", "", "Filter node list to those matching list"),
+        ("nodegroup", "", "Filter node list to those matching nodegroup"),
+        ("pillar", "", "Filter node list to those with matching pillar"),
+        ("regex", "", "Filter node list to those matching regex"),
+        ("subnet", "", "Filter node list to those in subnet"),
+    ]
+    targeted = False
+
+    def _get_up_nodes(self) -> list:
```

Comment:
What does "up" mean here in the context of a Saltstack cluster? Literally just on? Reachable on the network? Functioning within the cluster?

I ask because I want to make sure that we're able to get hostnames of nodes that are problematic and thus interesting to look into those sos reports. If a salt node is having issues and we want to collect a report from it and other nodes, I would prefer that we return that node as part of the list of nodes we're collecting from, and then rely on the other code flows to report we could not collect a report from it for XYZ reason.

Reply:
Responsive to the cluster (i.e. the salt-minion service is running, minion keys are accepted on the master, etc.). It does not denote the host is offline, because the service could simply be down. If the node is down however we will NOT be able to return its hostname because that is a grain which has to be queried via the minion, through the salt-minion service, which is not responsive on down nodes.

Reply:
If no minion ID is explicitly defined, it will default to the systems hostname. Either a fqdn or shortname depending on how the administrator configures the host.

My view on minion id being a hostname being uncommon may very well be skewed due to only seeing around 5-10% clusters I've worked on.  It could easily be the opposite is more common and that a minion ID is almost never explicitly defined and at some form of hostname is the minion id.

Reply:
I think it'd be preferable to try and return *something* for a "down" node that might still be reachable, since ostensibly we're trying to collect sos reports for a problematic condition on at least one host.

If what the cluster returns is not resolvable due to how the end user has configured things, then I don't think there's much we can be expected to do in that case - and we'll just handle the "invalid" hostnames like we do for truly down nodes - a "failed to connect" error which I would expect most end users to be able to piece together "oh, this minion ID isn't directly resolvable and that's why...".

Reply:
The last commit persists this behavior to ensure all nodes are part of a manifest. Regardless of up/down all nodes are present in the `manifest.json` list of nodes. The end user can work out whether one minion is not resolvable, or the node is down, etc.